File: backend/app/analysis/code_formatter.py
import re

def format_code_snippet(raw_code: str | None, language: str = 'python') -> str:
    """
    Pretty-prints and formats code snippets (Level 1, Level 2, etc.).
    Ensures consistent indentation, spacing, and block structure.
    """
    if not raw_code:
        return ''

    code = str(raw_code).strip()

    # 1. Strip Markdown Code Fences
    if code.startswith('```'):
        first_newline = code.find('\n')
        if first_newline != -1:
            code = code[first_newline + 1:]
        if code.endswith('```'):
            code = code[:-3]
        code = code.strip()

    # 2. Unescape literal string representations
    if '\\n' in code or '\\t' in code:
        code = code.replace('\\n', '\n').replace('\\t', '    ').replace('\\"', '"').replace("\\'", "'")

    norm_lang = (language or '').lower().strip()
    is_python = norm_lang in ('python', 'py')

    # 3. Pre-process lines for squished C/C++/Java/JS single-liners
    raw_lines: list[str] = []
    if not is_python and '\n' not in code and (';' in code or '{' in code):
        buffer = ''
        for char in code:
            buffer += char
            if char in (';', '{', '}'):
                raw_lines.append(buffer.strip())
                buffer = ''
        if buffer.strip():
            raw_lines.append(buffer.strip())
    else:
        raw_lines = code.split('\n')

    formatted_lines: list[str] = []
    indent_level = 0

    for raw_line in raw_lines:
        line = raw_line.strip()
        if not line:
            if formatted_lines and formatted_lines[-1] != '':
                formatted_lines.append('')
            continue

        # Spacing fixes around control keywords and braces
        line = re.sub(r'\b(if|for|while|switch|catch)\s*\(', r'\1 (', line)
        line = re.sub(r'\)\s*\{', ') {', line)
        line = re.sub(r'\belse\s*\{', 'else {', line)
        line = re.sub(r'\btry\s*\{', 'try {', line)
        line = re.sub(r'\bdo\s*\{', 'do {', line)

        if is_python:
            is_dedent = bool(re.match(r'^(elif\b|else\b|except\b|finally\b)', line))
            if is_dedent and indent_level > 0:
                indent_level = max(0, indent_level - 1)

            current_indent = '    ' * indent_level
            formatted_lines.append(current_indent + line)

            if line.endswith(':') or ': #' in line or ':#' in line:
                indent_level += 1
        else:
            open_braces = line.count('{')
            close_braces = line.count('}')

            if line.startswith('}'):
                indent_level = max(0, indent_level - 1)

            current_indent = '    ' * indent_level
            formatted_lines.append(current_indent + line)

            if not line.startswith('}'):
                net_change = open_braces - close_braces
                indent_level = max(0, indent_level + net_change)
            else:
                net_change = open_braces - (close_braces - 1)
                if net_change > 0:
                    indent_level += net_change

    while formatted_lines and formatted_lines[-1] == '':
        formatted_lines.pop()

    return '\n'.join(formatted_lines)

File: frontend/src/utils/codeFormatter.ts
/**
 * Automatically formats and pretty-prints code snippets for AI Review (Level 1, Level 2, etc.).
 * Ensures consistent indentation, proper newline splitting for squished statements,
 * correct spacing around keywords and braces, and clean block hierarchy.
 */

export const formatCodeSnippet = (rawCode: string | undefined | null, lang: string = 'python'): string => {
  if (!rawCode) return '';

  let code = String(rawCode).trim();

  // 1. Strip Markdown Code Fences (```python ... ``` or ``` ...)
  if (code.startsWith('```')) {
    const firstNewline = code.indexOf('\n');
    if (firstNewline !== -1) {
      code = code.slice(firstNewline + 1);
    }
    if (code.endsWith('```')) {
      code = code.slice(0, -3);
    }
    code = code.trim();
  }

  // 2. Convert literal escaped character sequences (\n, \t, \", \') to real characters
  if (code.includes('\\n') || code.includes('\\t')) {
    code = code
      .replace(/\\n/g, '\n')
      .replace(/\\t/g, '    ')
      .replace(/\\"/g, '"')
      .replace(/\\'/g, "'");
  }

  const normalizedLang = (lang || '').toLowerCase().trim();
  const isPython = normalizedLang === 'python' || normalizedLang === 'py';

  // 3. Pre-process lines: If C/C++/Java/JS code has multiple statements squished into one line using ';' or '{' or '}'
  let rawLines: string[] = [];

  if (!isPython && !code.includes('\n') && (code.includes(';') || code.includes('{'))) {
    let buffer = '';
    for (let i = 0; i < code.length; i++) {
      const char = code[i];
      buffer += char;
      if (char === ';' || char === '{' || char === '}') {
        rawLines.push(buffer.trim());
        buffer = '';
      }
    }
    if (buffer.trim()) {
      rawLines.push(buffer.trim());
    }
  } else {
    rawLines = code.split('\n');
  }

  // 4. Format lines with block-level indentation & keyword spacing
  const formattedLines: string[] = [];
  let indentLevel = 0;

  for (let idx = 0; idx < rawLines.length; idx++) {
    let line = rawLines[idx].trim();

    // Skip empty lines at the very top or consecutive blank lines
    if (!line) {
      if (formattedLines.length > 0 && formattedLines[formattedLines.length - 1] !== '') {
        formattedLines.push('');
      }
      continue;
    }

    // Keyword & brace spacing normalizations for C/C++/Java/JS/Python
    line = line
      .replace(/\b(if|for|while|switch|catch)\s*\(/g, '$1 (')
      .replace(/\)\s*\{/g, ') {')
      .replace(/\belse\s*\{/g, 'else {')
      .replace(/\btry\s*\{/g, 'try {')
      .replace(/\bdo\s*\{/g, 'do {')
      .replace(/\bfinally\s*\{/g, 'finally {');

    if (isPython) {
      // Python indentation logic
      const isDedentLine = /^(elif\b|else\b|except\b|finally\b)/.test(line);

      if (isDedentLine && indentLevel > 0) {
        indentLevel = Math.max(0, indentLevel - 1);
      }

      const currentIndent = '    '.repeat(indentLevel);
      formattedLines.push(currentIndent + line);

      if (line.endsWith(':') || line.includes(': #') || line.includes(':#')) {
        indentLevel++;
      }
    } else {
      // C / C++ / Java / JS indentation logic based on braces '{' and '}'
      const openBraces = (line.match(/\{/g) || []).length;
      const closeBraces = (line.match(/\}/g) || []).length;

      if (line.startsWith('}')) {
        indentLevel = Math.max(0, indentLevel - 1);
      }

      const currentIndent = '    '.repeat(indentLevel);
      formattedLines.push(currentIndent + line);

      if (!line.startsWith('}')) {
        const netChange = openBraces - closeBraces;
        indentLevel = Math.max(0, indentLevel + netChange);
      } else {
        const netChange = openBraces - (closeBraces - 1);
        if (netChange > 0) {
          indentLevel += netChange;
        }
      }
    }
  }

  // Remove trailing empty lines
  while (formattedLines.length > 0 && formattedLines[formattedLines.length - 1] === '') {
    formattedLines.pop();
  }

  return formattedLines.join('\n');
};
